### PR TITLE
pkg/authn/delegating: dynamically reload client CA

### PR DESCRIPTION
**Pull request by s-urbaniak**
cc @stlaz @deads2k

Signed-off-by: Sergiusz Urbaniak <sergiusz.urbaniak@gmail.com>

### CODE REVIEW

**Comment by s-urbaniak** [comment]
cc @brancz @paulfantom ptal if you have any free cycles as well :pray:

**Review by paulfantom — APPROVED** [review]
nice! lgtm :+1:

File: pkg/authn/delegating.go [hunk 1]
```diff
@@ -62,6 +55,6 @@ func NewDelegatingAuthenticator(client authenticationclient.TokenReviewInterface
 		APIAudiences:                       authenticator.Audiences(authn.Token.Audiences),
 	}
 
-	authenticator, _, err := authenticatorConfig.New()
-	return authenticator, err
+	auth, _, err := authenticatorConfig.New()
```
Comment (deads2k) [inline pkg/authn/delegating.go:58]: eh,  if I'm reviewing, I prefer the original.  We don't charge extra for letters and authenticator and authorizer both share auth.

File: pkg/authn/delegating.go [hunk 2]
```diff
@@ -29,29 +28,23 @@ import (
...
 		if err != nil {
 			return nil, err
 		}
 
-		p, err = dynamiccertificates.NewStaticCAContent(authn.X509.ClientCAFile, caBundle)
-		if err != nil {
-			return nil, err
-		}
+		// worker count is ignored
+		go p.Run(1, ctx.Done())
```
Comment (deads2k) [inline pkg/authn/delegating.go:47]: I recommend against having New methods start gofuncs.  Instead, you can return an interface that includes both authenticator.Request and a Run and start this in the caller.

You can call `RunOnce` in this method to ensure you have a valid authenticator to begin.

File: pkg/authn/delegating.go (hunk 1 above)
Comment (s-urbaniak) [reply pkg/authn/delegating.go:58]: this shadows the "authenticator" package import but i am agnostic as well.

File: pkg/authn/delegating.go (hunk 2 above)
Comment (s-urbaniak) [reply pkg/authn/delegating.go:47]: agreed, and i was having a similar thought, but didn't want to complicate things. agreed, let's return an enhanced interface and start outside the constructor :+1:

Comment (s-urbaniak) [reply pkg/authn/delegating.go:47]: > You can call RunOnce in this method to ensure you have a valid authenticator to begin.
I think we're fine here as the bundle is loaded in the constructor already https://github.com/kubernetes/apiserver/blob/e78fb1931d1a7682240456bf1883759720b97bb1/pkg/server/dynamiccertificates/dynamic_cafile_content.go#L85-L87

**Pull request merged**